optuna/study.py

```diff
@@ -13,26 +14,25 @@
 from typing import Tuple
 from typing import Type
 from typing import Union
-import warnings
 
 import joblib
-from joblib import delayed
 from joblib import Parallel
+from joblib import delayed
 
-from optuna._experimental import experimental
-from optuna._imports import try_import
-from optuna._study_direction import StudyDirection
-from optuna._study_summary import StudySummary  # NOQA
 from optuna import exceptions
 from optuna import logging
 from optuna import progress_bar as pbar_module
 from optuna import pruners
 from optuna import samplers
 from optuna import storages
 from optuna import trial as trial_module
-from optuna.trial import create_trial
+from optuna._experimental import experimental
+from optuna._imports import try_import
+from optuna._study_direction import StudyDirection
+from optuna._study_summary import StudySummary  # NOQA
```

Comment:
Because these imports are placed after the imports of pruners, samplers, storages, and trials, the circular imports occur. How about separating these imports and other optuna imports by a blank line? The specific change will resolve circular imports.

Reply:
Isort ignores a separation by a blank line.

![isort](https://user-images.githubusercontent.com/17039389/92595188-eb25b580-f2de-11ea-8940-9dd9ee15dfb2.gif)

Reply:
Hmm... Thanks for the explanation. So in order to use `isort` in Optuna, we have to allow this rule: "not `from optuna.study import StudyDirection` but `from optuna._study_direction import StudyDirection` in the Optuna codebase". But, int the `integration` module, we can use `from optuna.study import StudyDirection`?

We should note this rule in `CONTRIBUTING.md` as a follow-up task.

Reply:
or we can use `isort: skip` like:

```
# don't sort these imports
import a  # isort: skip
import b  # isort: skip
import c  # isort: skip

import ...
import ...
```

Reply:
Thanks for providing an important suggestion. Looks good. I think it's better to use `isort: skip` for `from optuna._study_direction import StudyDirection` and `from optuna._study_summary import StudySummary`.  What do you think?

Reply:
I internally discussed with Optuna core-dev members. As a result, we can accept `from optuna._study_direction import StudyDirection` in Optuna codebase. This new convention leads to reduce the number of codes dependent on the order of imports.

As a follow-up task, we should note in `CONTRIBUTING.md` to use `from optuna._study_direction import StudyDirection`.

Reply:
Makes sense.